servers/src/common/adapters.rs

```diff
@@ -610,8 +643,45 @@ impl NetToChainAdapter {
 
 		let bhash = b.hash();
 		let previous = self.chain().get_previous_header(&b.header);
+		let within_checkpointed_range;
+		let mut disable_checkpoints = false;
+		let mut options = opts;
+
+		let network_height = self.chain().header_head()?.height;
+		// ensure we check proof-of-work for last (POW_VERIFICATION_THRESHOLD) blocks from network chaintip
+		let check_pow_dyn_threshold = network_height - POW_VERIFICATION_THRESHOLD;
+
+		if self.config.disable_checkpoints.is_some() {
+			if self.config.disable_checkpoints.unwrap() {
+				// don't honor disable_checkpoints once we hit dynamic threshold
+				if b.header.height < check_pow_dyn_threshold {
+					disable_checkpoints = true;
+				}
+			}
+		}
+
+		match self.chain().check_header_against_checkpoints(&b.header) {
+			Ok(in_range) => {
+				within_checkpointed_range = in_range;
+			}
+			Err(e) => {
+				return Err(e);
+			}
+		}
+
+		if self.config.skip_pow_validation.is_some() {
+			if self.config.skip_pow_validation.unwrap() {
+				if within_checkpointed_range || disable_checkpoints {
+					// skip pow validation if we are within checkpointed range OR
+					// if we have disable_checkpoints = true AND we are outside of
+					// dynamic chaintip threshold, otherwise fully verify
+					options = chain::Options::SKIP_POW;
+				}
+				//warn!("b.header.height({}), dyn_threshold({}), options({:?})", b.header.height, check_pow_dyn_threshold, options);
+			}
+		}
 
-		match self.chain().process_block(b, opts) {
+		match self.chain().process_block(b, options) {
```

Comment:
@who-biz i have still concerns about this skip_pow option in this process_block function. For header_sync the option to disable pow check is fine, but on blocks?

Reply:
If you'd like to remove it, we can! I doubt performance would suffer that much since we really aren't syncing that many blocks after `HeaderSync`.

The last 300 (`MAX_ORPHAN_SIZE * 5`) blocks are always verified for pow. `disable_checkpoints` config setting actually has no effect here on blocks within that threshold. (See here: https://github.com/EpicCash/epic/pull/105/files#diff-172148fdc6d0a8a940347d1429aa3519410b304d29d8c81a8c9544b5bd2dcf31R650-R670)

But, if you'd like to fully verify all PoW during `BodySync`, we can do that.

Reply:
I can also add a separate `bool` here to make this behavior more readable...

Something like changing:
```
if within_checkpointed_range || disable_checkpoints
```
To something more closely resembling:
```
if within_checkpointed_range || (disable_checkpoints && outside_forced_pow_threshold)
```

Reply:
Hi, ok did not see that skip pow is only for older blocks.

This is better to read:
if within_checkpointed_range || (disable_checkpoints && outside_forced_pow_threshold

Reply:
Resolved in: https://github.com/EpicCash/epic/pull/105/commits/dd07ba0b76110c47e54de3e64896ce859a4d7966